cpp/src/toppra/geometric_path/piecewise_poly_path.cpp

```diff
@@ -28,6 +29,86 @@ PiecewisePolyPath::PiecewisePolyPath(const Matrices & coefficients,
   computeDerivativesCoefficients();
 }
 
+PiecewisePolyPath::PiecewisePolyPath(const Vectors &positions, const Vector &times,
+        const std::array<BoundaryCond, 2> &bc_type) {
+    // Prepare input
+    checkInputArgs(positions, times, bc_type);
+
+    // h(i) = t(i+1) - t(i)
+    Vector h (times.rows() - 1);
+    for (size_t i = 0; i < h.rows(); i++){
+        h(i) = times(i + 1) - times(i);
+    }
+
+    // Construct the tri-diagonal matrix A based on spline continuity criteria
+    Matrix A = Matrix::Zero(times.rows(), times.rows());
+    for (size_t i = 1; i < A.rows() - 1; i++) {
+        A.row(i).segment(i - 1, 3) << h(i - 1), 2 * (h(i - 1) + h(i)), h(i);
+    }
+
+    // Construct B based on spline continuity criteria
+    Vectors B (positions.at(0).rows());
+    for (size_t i = 0; i < B.size(); i++) {
+        B[i].resize(times.rows());
+        for (size_t j = 1; j < A.rows() - 1; j++) {
+            B[i](j) = 3 * (positions[j + 1](i) - positions[j](i)) / h(j) -
+                    3 * (positions[j](i) - positions[j - 1](i)) / h(j - 1);
+        }
+    }
+
+    // Insert boundary conditions to A and B
+    if (bc_type[0].order == 1) {
+        A.row(0).segment(0, 2) << 2 * h(0), h(0);
+        for (size_t i = 0; i < B.size(); i++) {
+            B[i](0) = 3 * (positions[1](i) - positions[0](i)) / h(0) - 3 * bc_type[0].values(i);
+        }
+    }
+    else if (bc_type[0].order == 2) {
+        A(0, 0) = 2;
+        for (size_t i = 0; i < B.size(); i++) {
+            B[i](0) = bc_type[0].values(i);
+        }
+    }
+
+    if (bc_type[1].order == 1) {
+        A.row(A.rows() - 1).segment(A.cols() - 2, 2) << h(h.rows() - 1), 2 * h(h.rows() - 1);
+        for (size_t i = 0; i < B.size(); i++) {
+            B[i](B[i].rows() - 1) =
+                    3 * bc_type[1].values(i) -
+                    3 * (positions[positions.size() - 1](i) - positions[positions.size() - 2](i)) / h(h.rows() - 1);
+        }
+    }
+    else if (bc_type[1].order == 2) {
+        A(A.rows() - 1, A.cols() - 1) = 2;
+        for (size_t i = 0; i < B.size(); i++) {
+            B[i](B[i].rows() - 1) = bc_type[1].values(i);
+        }
+    }
+
+    // Solve AX = B
+    Vectors X (positions[0].rows());
+    for (size_t i = 0; i < X.size(); i++) {
+        X[i].resize(times.rows());
+        X[i] = A.colPivHouseholderQr().solve(B[i]);
+    }
+
+    // Insert spline coefficients
+    Matrices coefficients(times.size() - 1);
+    for (size_t i = 0; i < coefficients.size() ; i++) {
+        coefficients[i].resize(4, positions[0].rows());
+        for (size_t j = 0; j < coefficients[i].cols(); j++) {
+            coefficients[i](0, j) = (X[j](i + 1) - X[j](i)) / (3 * h(i));
+            coefficients[i](1, j) = X[j](i);
+            coefficients[i](2, j) = (positions[i + 1](j) - positions[i](j)) / h(i) -
+                    h(i) / 3 * (2 * X[j](i) + X[j](i + 1));
+            coefficients[i](3, j) = positions[i](j);
+        }
+    }
+
```

Comment:
I think this portion can be extracted to a separate function for readability. It will also make future extension easier.

Reply:
Noted. Changed in https://github.com/hungpham2511/toppra/pull/161/commits/9023de3c13e92777776f6dd9b0b05fb78f2fbe84. Thanks.